File: Frontend/DesktopApp/src/main/java/com/spam/financialaccounting/desktop/api/ApiClient.java
package com.spam.financialaccounting.desktop.api;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.spam.financialaccounting.desktop.config.AppConfig;
import com.spam.financialaccounting.desktop.model.BalanceSheet;
import com.spam.financialaccounting.desktop.model.FAGroup;
import com.spam.financialaccounting.desktop.model.FASubGroup;
import com.spam.financialaccounting.desktop.model.JournalDetail;
import com.spam.financialaccounting.desktop.model.JournalMaster;
import com.spam.financialaccounting.desktop.model.PeriodLock;
import com.spam.financialaccounting.desktop.model.ProfitAndLoss;
import com.spam.financialaccounting.desktop.model.TrialBalance;

public class ApiClient {
    private static final String BASE_URL = AppConfig.getBaseUrl();

    private final HttpClient client;
    private final ObjectMapper mapper;

    public ApiClient() {
        this.client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(5))
                .build();
        this.mapper = new ObjectMapper();
        this.mapper.registerModule(new JavaTimeModule());
        this.mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
        // Write dates as ISO-8601 strings (e.g. "2026-06-08T00:00:00"), not numeric arrays
        this.mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
    }

    public CompletableFuture<Boolean> pingAsync() {
        HttpRequest request = HttpRequest.newBuilder().uri(URI.create(BASE_URL + "/fagroups"))
                .timeout(Duration.ofSeconds(3)).GET().build();

        return client.sendAsync(request, HttpResponse.BodyHandlers.discarding())
                .thenApply(response -> response.statusCode() == 200).exceptionally(ex -> false);
    }

    // FAGROUP API
    public List<FAGroup> getGroups() throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/fagroups")).GET().build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), new TypeReference<List<FAGroup>>() {
        });
    }

    public FAGroup createGroup(FAGroup group) throws IOException, InterruptedException {
        String json = mapper.writeValueAsString(group);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/fagroups"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), FAGroup.class);
    }

    public FAGroup updateGroup(String code, FAGroup group) throws IOException, InterruptedException {
        String json = mapper.writeValueAsString(group);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/fagroups/" + code))
                .header("Content-Type", "application/json")
                .PUT(HttpRequest.BodyPublishers.ofString(json))
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), FAGroup.class);
    }

    // --- FASUBGROUP API (Ledgers) ---
    public List<FASubGroup> getLedgerAccounts() throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/ledger-accounts"))
                .GET()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), new TypeReference<List<FASubGroup>>() {
        });
    }

    public FASubGroup createLedgerAccount(FASubGroup subGroup) throws IOException, InterruptedException {
        String json = mapper.writeValueAsString(subGroup);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/ledger-accounts"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), FASubGroup.class);
    }

    public FASubGroup updateLedgerAccount(String code, FASubGroup subGroup) throws IOException, InterruptedException {
        String json = mapper.writeValueAsString(subGroup);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/ledger-accounts/" + code))
                .header("Content-Type", "application/json")
                .PUT(HttpRequest.BodyPublishers.ofString(json))
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), FASubGroup.class);
    }

    public void deleteLedgerAccount(String code) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/ledger-accounts/" + code))
                .DELETE()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
    }

    // --- JOURNAL MASTER ---
    public List<JournalMaster> getJournalMasters() throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/journal-masters"))
                .GET()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), new TypeReference<List<JournalMaster>>() {
        });
    }

    public JournalMaster createJournalMaster(JournalMaster master) throws IOException, InterruptedException {
        String json = mapper.writeValueAsString(master);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/journal-masters"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), JournalMaster.class);
    }

    public void deleteJournalMaster(String jId) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/journal-masters/" + jId))
                .DELETE()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
    }

    // Journal detail
    public List<JournalDetail> getJournalDetailsByJournalId(String jId) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/journal-details/journal/" + jId))
                .GET()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), new TypeReference<List<JournalDetail>>() {
        });
    }

    public JournalDetail createJournalDetail(JournalDetail detail) throws IOException, InterruptedException {
        String json = mapper.writeValueAsString(detail);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/journal-details"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), JournalDetail.class);
    }

    public void deleteJournalDetail(String jId, String jCode, String jDrCr) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/journal-details/" + jId + "/" + jCode + "/" + jDrCr))
                .DELETE()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
    }

    // --- JOURNAL VOUCHERS (balanced; post/reverse/delete happen atomically) ---
    public JournalMaster postJournalVoucher(Object request) throws IOException, InterruptedException {
        String json = mapper.writeValueAsString(request);
        HttpRequest httpRequest = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/journal-vouchers"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();
        HttpResponse<String> response = client.send(httpRequest, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), JournalMaster.class);
    }

    public void deleteJournalVoucher(String jId) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/journal-vouchers/" + jId))
                .DELETE()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
    }

    public JournalMaster reverseJournalVoucher(String jId) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/journal-vouchers/" + jId + "/reverse"))
                .POST(HttpRequest.BodyPublishers.noBody())
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), JournalMaster.class);
    }

    // --- OPENING BALANCES ---
    public JournalMaster importOpeningBalances(Object request) throws IOException, InterruptedException {
        String json = mapper.writeValueAsString(request);
        HttpRequest httpRequest = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/opening-balances/import"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();
        HttpResponse<String> response = client.send(httpRequest, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), JournalMaster.class);
    }

    // --- PERIOD LOCKS ---
    public List<PeriodLock> getPeriodLocks() throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/period-locks"))
                .GET()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), new TypeReference<List<PeriodLock>>() {
        });
    }

    public PeriodLock createPeriodLock(Object request) throws IOException, InterruptedException {
        String json = mapper.writeValueAsString(request);
        HttpRequest httpRequest = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/period-locks"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();
        HttpResponse<String> response = client.send(httpRequest, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), PeriodLock.class);
    }

    public void deletePeriodLock(long id) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/period-locks/" + id))
                .DELETE()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
    }

    // --- REPORTS ---
    // asOfDate (yyyy-MM-dd) limits the report to vouchers posted on or before that
    // date. Pass null to report as of today.
    public TrialBalance getTrialBalance(String asOfDate) throws IOException, InterruptedException {
        String query = (asOfDate != null && !asOfDate.isBlank()) ? "?asOfDate=" + asOfDate : "";
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/reports/trial-balance" + query))
                .GET()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), TrialBalance.class);
    }

    public ProfitAndLoss getProfitAndLoss(String asOfDate) throws IOException, InterruptedException {
        String query = (asOfDate != null && !asOfDate.isBlank()) ? "?asOfDate=" + asOfDate : "";
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/reports/profit-and-loss" + query))
                .GET()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), ProfitAndLoss.class);
    }

    public BalanceSheet getBalanceSheet(String asOfDate) throws IOException, InterruptedException {
        String query = (asOfDate != null && !asOfDate.isBlank()) ? "?asOfDate=" + asOfDate : "";
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/reports/balance-sheet" + query))
                .GET()
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        handleErrorResponse(response);
        return mapper.readValue(response.body(), BalanceSheet.class);
    }

    private void handleErrorResponse(HttpResponse<String> response) throws IOException {
        if (response.statusCode() >= 400) {
            String errorMsg = "HTTP Status Code " + response.statusCode();
            try {
                // pull the message out of Spring Boot's standard error body
                var map = mapper.readValue(response.body(), new TypeReference<Map<String, Object>>() {
                });
                if (map.containsKey("message")) {
                    errorMsg = map.get("message").toString();
                } else if (map.containsKey("error")) {
                    errorMsg = map.get("error").toString();
                }
            } catch (Exception ignored) {}
            throw new IOException(errorMsg);
        }
    }
}

File: Frontend/web-app/src/api/apiClient.ts
import type {
  BalanceSheet,
  FAGroup,
  FASubGroup,
  JournalDetail,
  JournalMaster,
  JournalVoucherRequest,
  OpeningBalanceImportRequest,
  PeriodLock,
  PeriodLockRequest,
  ProfitAndLoss,
  TrialBalance,
} from "../types";

const BASE_URL = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8080/api/v1";

/** Timeout for API calls (8 seconds) */
const DEFAULT_TIMEOUT_MS = 8000;

function makeSignal(): AbortSignal {
  return AbortSignal.timeout(DEFAULT_TIMEOUT_MS);
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    let errMsg = `HTTP Error ${response.status}`;
    try {
      const data = await response.json();
      errMsg = data.message || data.error || errMsg;
    } catch {
      // body might not be JSON, ignore
    }
    throw new Error(errMsg);
  }

  if (response.status === 204) return null as T;

  return response.json();
}

export const apiClient = {
  //  Connection Health 
  async ping(): Promise<boolean> {
    try {
      const res = await fetch(`${BASE_URL}/fagroups`, {
        method: "GET",
        signal: AbortSignal.timeout(3000),
      });
      return res.ok;
    } catch {
      return false;
    }
  },

  // Account Groups
  async getGroups(): Promise<FAGroup[]> {
    return handleResponse<FAGroup[]>(
      await fetch(`${BASE_URL}/fagroups`, { signal: makeSignal() }),
    );
  },

  async createGroup(group: FAGroup): Promise<FAGroup> {
    return handleResponse<FAGroup>(
      await fetch(`${BASE_URL}/fagroups`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(group),
        signal: makeSignal(),
      }),
    );
  },

  async updateGroup(code: string, group: FAGroup): Promise<FAGroup> {
    return handleResponse<FAGroup>(
      await fetch(`${BASE_URL}/fagroups/${code}`, {
        method: "PUT",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(group),
        signal: makeSignal(),
      }),
    );
  },

  async deleteGroup(code: string): Promise<void> {
    return handleResponse<void>(
      await fetch(`${BASE_URL}/fagroups/${code}`, {
        method: "DELETE",
        signal: makeSignal(),
      }),
    );
  },

  //  Ledger Accounts (Sub-Groups) 
  async getLedgers(): Promise<FASubGroup[]> {
    return handleResponse<FASubGroup[]>(
      await fetch(`${BASE_URL}/ledger-accounts`, { signal: makeSignal() }),
    );
  },

  async createLedger(ledger: FASubGroup): Promise<FASubGroup> {
    return handleResponse<FASubGroup>(
      await fetch(`${BASE_URL}/ledger-accounts`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(ledger),
        signal: makeSignal(),
      }),
    );
  },

  async updateLedger(code: string, ledger: FASubGroup): Promise<FASubGroup> {
    return handleResponse<FASubGroup>(
      await fetch(`${BASE_URL}/ledger-accounts/${code}`, {
        method: "PUT",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(ledger),
        signal: makeSignal(),
      }),
    );
  },

  async deleteLedger(code: string): Promise<void> {
    return handleResponse<void>(
      await fetch(`${BASE_URL}/ledger-accounts/${code}`, {
        method: "DELETE",
        signal: makeSignal(),
      }),
    );
  },

  //  Journal Masters (read-only; writes go through Journal Vouchers)
  async getJournalMasters(): Promise<JournalMaster[]> {
    return handleResponse<JournalMaster[]>(
      await fetch(`${BASE_URL}/journal-masters`, { signal: makeSignal() }),
    );
  },

  // Journal Details (read-only; writes go through Journal Vouchers)
  async getJournalDetails(jId: string): Promise<JournalDetail[]> {
    return handleResponse<JournalDetail[]>(
      await fetch(`${BASE_URL}/journal-details/journal/${jId}`, {
        signal: makeSignal(),
      }),
    );
  },

  //  Journal Vouchers (balanced, atomic post/reverse/delete)
  async postJournalVoucher(
    voucher: JournalVoucherRequest,
  ): Promise<JournalMaster> {
    return handleResponse<JournalMaster>(
      await fetch(`${BASE_URL}/journal-vouchers`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(voucher),
        signal: makeSignal(),
      }),
    );
  },

  async deleteJournalVoucher(jId: string): Promise<void> {
    return handleResponse<void>(
      await fetch(`${BASE_URL}/journal-vouchers/${jId}`, {
        method: "DELETE",
        signal: makeSignal(),
      }),
    );
  },

  // Posts a reversing voucher that cancels the given one and returns the new reversal header.
  async reverseJournalVoucher(jId: string): Promise<JournalMaster> {
    return handleResponse<JournalMaster>(
      await fetch(`${BASE_URL}/journal-vouchers/${jId}/reverse`, {
        method: "POST",
        signal: makeSignal(),
      }),
    );
  },

  //  Opening Balances (imported as one atomic Opening Balance voucher)
  async importOpeningBalances(
    request: OpeningBalanceImportRequest,
  ): Promise<JournalMaster> {
    return handleResponse<JournalMaster>(
      await fetch(`${BASE_URL}/opening-balances/import`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(request),
        signal: makeSignal(),
      }),
    );
  },

  //  Period Locks (closed accounting periods)
  async getPeriodLocks(): Promise<PeriodLock[]> {
    return handleResponse<PeriodLock[]>(
      await fetch(`${BASE_URL}/period-locks`, { signal: makeSignal() }),
    );
  },

  async createPeriodLock(request: PeriodLockRequest): Promise<PeriodLock> {
    return handleResponse<PeriodLock>(
      await fetch(`${BASE_URL}/period-locks`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(request),
        signal: makeSignal(),
      }),
    );
  },

  async deletePeriodLock(id: number): Promise<void> {
    return handleResponse<void>(
      await fetch(`${BASE_URL}/period-locks/${id}`, {
        method: "DELETE",
        signal: makeSignal(),
      }),
    );
  },

  //  Reports
  // asOfDate (ISO yyyy-MM-dd) limits the report to vouchers posted on or before
  // that date. Leave it off to report as of today.
  async getTrialBalance(asOfDate?: string): Promise<TrialBalance> {
    const query = asOfDate ? `?asOfDate=${encodeURIComponent(asOfDate)}` : '';
    return handleResponse<TrialBalance>(
      await fetch(`${BASE_URL}/reports/trial-balance${query}`, { signal: makeSignal() }),
    );
  },

  async getProfitAndLoss(asOfDate?: string): Promise<ProfitAndLoss> {
    const query = asOfDate ? `?asOfDate=${encodeURIComponent(asOfDate)}` : '';
    return handleResponse<ProfitAndLoss>(
      await fetch(`${BASE_URL}/reports/profit-and-loss${query}`, { signal: makeSignal() }),
    );
  },

  async getBalanceSheet(asOfDate?: string): Promise<BalanceSheet> {
    const query = asOfDate ? `?asOfDate=${encodeURIComponent(asOfDate)}` : '';
    return handleResponse<BalanceSheet>(
      await fetch(`${BASE_URL}/reports/balance-sheet${query}`, { signal: makeSignal() }),
    );
  },
};